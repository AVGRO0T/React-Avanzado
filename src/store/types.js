export const AUTH_LOGIN_REQUEST = 'AUTH_LOGIN_REQUEST';
export const AUTH_LOGIN_SUCCESS = 'AUTH_LOGIN_SUCCESS';
export const AUTH_LOGIN_FAILURE = 'AUTH_LOGIN_FAILURE';

export const AUTH_LOGOUT_REQUEST = 'AUTH_LOGOUT_REQUEST';
export const AUTH_LOGOUT_SUCCESS = 'AUTH_LOGOUT_SUCCESS';
export const AUTH_LOGOUT_FAILURE = 'AUTH_LOGOUT_FAILURE';

export const ADVERTS_LOADED_SUCCESS ='ADVERTS_LOADED_SUCCESS';
export const ADVERTS_LOADED_FAILURE ='ADVERTS_LOADED_FAILURE';
export const ADVERTS_LOADED_REQUEST ='ADVERTS_LOADED_REQUEST';

export const FILTERS_LOADED_REQUEST = 'FILTERS_LOADED_REQUEST';
export const FILTERS_LOADED_FAILURE = 'FILTERS_LOADED_FAILURE';
export const FILTERS_LOADED_SUCCESS = 'FILTERS_LOADED_SUCCESS';

export const UI_RESET_ERROR= 'UI_RESET_ERROR';