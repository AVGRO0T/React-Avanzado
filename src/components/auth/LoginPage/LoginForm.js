import React from 'react';
import T from 'prop-types';

const canSubmit = ({ email, password }) =>
  [
    // valid email
    !!email,
    // valid password
    !!password,
  ].every(validation => validation); // all validations pass

function LoginForm({ onSubmit }) {
  const [credentials, setCredentials] = React.useState({
    email: '',
    password: '',
    remember: false,
  });
  const { email, password, remember } = credentials;

  const handleChange = ev => {
    setCredentials(oldCredentials => ({
      ...oldCredentials,
      [ev.target.name]:
        ev.target.type === 'checkbox' ? ev.target.checked : ev.target.value,
    }));
  };

  const handleSubmit = ev => {
    ev.preventDefault();
    onSubmit(credentials);
  };

  return (
    <form onSubmit={handleSubmit}>
      <input name="email" value={email} onChange={handleChange} />
      <input
        type="password"
        name="password"
        value={password}
        onChange={handleChange}
      />
      <input
        type="checkbox"
        name="remember"
        checked={remember}
        onChange={handleChange}
      />
      <button disabled={!canSubmit(credentials)}>Login</button>
    </form>
  );
}

LoginForm.propTypes = {
  onSubmit: T.func.isRequired,
};

export default LoginForm;
