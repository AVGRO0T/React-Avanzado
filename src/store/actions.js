import {AUTH_LOGIN_REQUEST, 
    AUTH_LOGIN_SUCCESS,
    AUTH_LOGIN_FAILURE,
    AUTH_LOGOUT_SUCCESS,
    UI_RESET_ERROR,
    ADVERTS_LOADED_SUCCESS,
    ADVERTS_LOADED_REQUEST,
    ADVERTS_LOADED_FAILURE,
    AUTH_LOGOUT_REQUEST,
    AUTH_LOGOUT_FAILURE,
} from './types';
//AUTH  LOGIN ACTIONS 
export const authLoginSuccess = () => ({
    type: AUTH_LOGIN_SUCCESS,
  });

export const authLoginRequest = () => ({
    type: AUTH_LOGIN_REQUEST,
   });
  
export const authLoginFailure = error => ({
    type: AUTH_LOGIN_FAILURE,
    payload: error,
    error: true,
  });
  
  export const authLogin = credentials => {
    return async function (dispatch, getState, { api }) {
      try {
        dispatch(authLoginRequest());
        await api.auth.login(credentials);
        dispatch(authLoginSuccess());
      } catch (error) {
        dispatch(authLoginFailure(error));
        throw error;
      }
    };
  };
//AUTH LOGOUT ATIONS
export const authLogoutSuccess = () => ({
    type: AUTH_LOGOUT_SUCCESS,
  });

export const authLogout = () => {
    return async function (dispatch, getState, { api }) {
      await api.auth.logout();
      dispatch(authLogoutSuccess());
    };
  };

// UI ACTIONS
  export const uiResetError = () => ({
    type: UI_RESET_ERROR,
  });

  
