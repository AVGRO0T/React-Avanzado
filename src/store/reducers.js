import {AUTH_LOGIN_REQUEST, 
        AUTH_LOGIN_SUCCESS,
        AUTH_LOGIN_FAILURE,
        AUTH_LOGOUT_FAILURE,
        AUTH_LOGOUT_REQUEST,
        AUTH_LOGOUT_SUCCESS,
        UI_RESET_ERROR,
} from './types';



const defaultState = {
    auth: false,
    ui: {
      isLoading: false,
      error: null,
    },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT_SUCCESS:
      return false;
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return {
      isLoading: false,
      error: action.payload,
    };
  }
  if (/_REQUEST$/.test(action.type)) {
    return {
      error: null,
      isLoading: true,
    };
  }
  if (/_SUCCESS$/.test(action.type)) {
    return {
      error: null,
      isLoading: false,
    };
  }
  if (action.type === UI_RESET_ERROR) {
    return {
      ...state,
      error: null,
    };
  }
  return state;
}