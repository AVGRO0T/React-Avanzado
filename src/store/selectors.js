export const getIsLogged = state => state.auth;

export const getUi = state => state.ui;



export const getAdvertsSelector = state => 
state.adverts.areLoaded ? state.adverts.data : [];

export const areLoadedAdverts = state => state.adverts.areLoaded;

export const areLoadedTags = state => state.tags.areLoaded;


export const getTagsSelector = state => 
state.tags.areLoaded ? state.tags.data : [];

export const getAdvert = advertId => state =>
  state.advert.data.find(advert => advert.id.toString() === advertId);