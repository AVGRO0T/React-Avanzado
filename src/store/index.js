import { createStore, combineReducers/* , applyMiddleware */ } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
//import thunk from 'redux-thunk';

import * as reducers from './reducers';

// const reducer = combineReducers({ auth, tweets });
const reducer = combineReducers(reducers);

export default function configureStore(preloadedState) {
  const store = createStore(reducer, preloadedState, composeWithDevTools());

  return store;}