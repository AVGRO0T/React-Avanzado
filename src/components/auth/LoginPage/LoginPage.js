import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

import { useAuthContext } from '../context';

import LoginForm from './LoginForm';

import { useDispatch, useSelector } from 'react-redux';
import { authLogin, uiResetError } from '../../../store/actions';
import { getUi } from '../../../store/selectors';

function LoginPage() {
  const navigate = useNavigate();
  const location = useLocation();

 
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);
 const resetErrorr = () => dispatch(uiResetError());
  const handleSubmit = async credentials =>  {
   dispatch(authLogin(credentials))
  
     
  };

  return (
    <div>
      <LoginForm onSubmit={handleSubmit} isLoading={isLoading} />
      {isLoading && <p>...login in nodepop</p>}
      {error && (
        <div onClick={resetErrorr} style={{ color: 'red' }}>
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
