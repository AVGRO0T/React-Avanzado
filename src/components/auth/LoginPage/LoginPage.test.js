import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { defaultState } from '../../../store/reducers'
import LoginPage from './LoginPage'
import { authLogin } from '../../../store/actions'

jest.mock('../../../store/actions');

describe('LoginPage', () => {
    
        
        const store = {
            getState:()=>defaultState,
            dispatch: jest.fn(),
            subscribe: () => {},
        }
        const renderComponent = () =>
        render(
          <Provider store={store}>
            <LoginPage />
          </Provider>,
        );
    
      test('Creando una Snapshot', () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
      });
      /* test('should dispatch authLogin action', () => {
        const credentials = {
            email: 'nicolas@paycom.es',
            password: 'string',
            remember: false,
        }
       renderComponent();
        
        expect(authLogin).toHaveBeenLastCalledWith({email:'nicolas',password:'string',remeber:false});
      }); */
});