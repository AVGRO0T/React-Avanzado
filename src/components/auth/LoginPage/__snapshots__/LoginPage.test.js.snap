// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div>
    <form>
      <input
        name="email"
        value=""
      />
      <input
        name="password"
        type="password"
        value=""
      />
      <input
        name="remember"
        type="checkbox"
      />
      <button
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;

exports[`LoginPage Creando un snapshot 1`] = `
<div>
  <div>
    <form>
      <input
        name="email"
        value=""
      />
      <input
        name="password"
        type="password"
        value=""
      />
      <input
        name="remember"
        type="checkbox"
      />
      <button
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;

exports[`LoginPage Creando una Snapshot 1`] = `
<div>
  <div>
    <form>
      <input
        name="email"
        value=""
      />
      <input
        name="password"
        type="password"
        value=""
      />
      <input
        name="remember"
        type="checkbox"
      />
      <button
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
